File: data-structures-array-processing-enhanced/src/algorithms/brute-force-solutions.ts
```typescript
/**
 * @fileoverview Brute-force or less optimal solutions for array manipulation problems.
 * These implementations are provided to illustrate alternative (often less efficient)
 * approaches and highlight the benefits of optimization.
 */

// Problem 1: Rotate Array (Brute Force / Auxiliary Array)

/**
 * Rotates an array to the right by k steps using an auxiliary array.
 * This is a straightforward approach but uses O(N) extra space.
 *
 * The algorithm works by creating a new array, placing elements from `nums`
 * into their rotated positions, and then copying the new array back to `nums`.
 *
 * @param nums The array of numbers to rotate. Modified in-place.
 * @param k The number of steps to rotate the array to the right.
 * @returns void (modifies nums in-place).
 *
 * Time Complexity: O(N), where N is the number of elements in `nums`.
 *   - One pass to populate the new array, one pass to copy back.
 * Space Complexity: O(N)
 *   - An auxiliary array of size N is created.
 */
export function rotateArrayBruteForce(nums: number[], k: number): void {
    const n = nums.length;
    if (n === 0 || k === 0) {
        return;
    }

    k %= n; // Handle cases where k > n

    if (k === 0) {
        return;
    }

    const rotated = new Array(n);
    for (let i = 0; i < n; i++) {
        // Calculate the new position for each element.
        // (i + k) % n gives the new index after rotating right by k steps.
        rotated[(i + k) % n] = nums[i];
    }

    // Copy elements back to the original array
    for (let i = 0; i < n; i++) {
        nums[i] = rotated[i];
    }
}

// Problem 2: Product of Array Except Self (Brute Force - using division implicitly)

/**
 * Calculates the product of all elements except self for each element.
 * This version uses division and handles zero values carefully.
 * This is generally not the preferred solution for interview problems that
 * explicitly forbid division, but it's an alternative.
 *
 * The algorithm calculates the total product of the array.
 * Then, for each element, it divides the total product by that element.
 * Special handling is required for zeros:
 * - If there are two or more zeros, all results will be zero.
 * - If there is exactly one zero, only the element at the zero's index will
 *   have a non-zero product (product of all other numbers).
 * - If no zeros, standard division applies.
 *
 * @param nums The input array of integers.
 * @returns An array where `answer[i]` is the product of all elements in `nums` except `nums[i]`.
 *
 * Time Complexity: O(N), where N is the number of elements in `nums`.
 *   - One pass to find total product and count zeros.
 *   - One pass to compute results.
 * Space Complexity: O(1) (excluding the output array).
 */
export function productExceptSelfBruteForceWithDivision(nums: number[]): number[] {
    const n = nums.length;
    const answer: number[] = new Array(n).fill(0); // Initialize with 0 for safety

    if (n === 0) {
        return [];
    }
    if (n === 1) {
        return [1];
    }

    let totalProduct = 1;
    let zeroCount = 0;
    let indexOfSingleZero = -1;

    // First pass: calculate total product and count zeros
    for (let i = 0; i < n; i++) {
        if (nums[i] === 0) {
            zeroCount++;
            indexOfSingleZero = i;
            continue; // Skip multiplying by zero to avoid making totalProduct 0 prematurely
        }
        totalProduct *= nums[i];
    }

    // Second pass: fill the answer array
    if (zeroCount > 1) {
        // If two or more zeros, all products will be 0
        return answer; // Already filled with zeros
    } else if (zeroCount === 1) {
        // If exactly one zero, only that position will have a non-zero product
        answer[indexOfSingleZero] = totalProduct; // totalProduct here is product of non-zero elements
        return answer; // Other positions remain 0
    } else {
        // No zeros, regular division
        for (let i = 0; i < n; i++) {
            answer[i] = totalProduct / nums[i];
        }
    }

    return answer;
}


// Problem 3: Merge Intervals (Brute Force - O(N^2) comparison)

/**
 * Merges overlapping intervals using a brute-force approach.
 * This method involves repeatedly iterating and merging until no more merges are possible.
 * It's less efficient than the sort-based approach.
 *
 * @param intervals An array of intervals, where each interval is [start, end].
 * @returns An array of non-overlapping intervals.
 *
 * Time Complexity: O(N^2) in the worst case, where N is the number of intervals.
 *   - In each pass, we might find one merge. If there are many overlaps, multiple passes
 *     might be needed. Each pass iterates N elements.
 *   - E.g., [[1,10],[2,3],[4,5],[6,7]] would require several passes.
 * Space Complexity: O(N) for the new array each time we merge.
 */
export function mergeIntervalsBruteForce(intervals: number[][]): number[][] {
    if (intervals.length <= 1) {
        return intervals;
    }

    let merged = false;
    // We need to keep iterating as long as we make merges
    // because merging [1,5] and [4,8] creates [1,8], which might then merge with [7,10]
    // which was not adjacent to [1,5] initially.
    do {
        merged = false;
        const newIntervals: number[][] = [];
        const used = new Array(intervals.length).fill(false);

        for (let i = 0; i < intervals.length; i++) {
            if (used[i]) continue;

            let current = intervals[i];
            let foundOverlap = false;

            for (let j = i + 1; j < intervals.length; j++) {
                if (used[j]) continue;

                const next = intervals[j];

                // Check for overlap: [a,b] and [c,d] overlap if a <= d AND c <= b
                if (Math.max(current[0], next[0]) <= Math.min(current[1], next[1])) {
                    // Overlap found, merge them
                    current = [Math.min(current[0], next[0]), Math.max(current[1], next[1])];
                    used[j] = true; // Mark next interval as used
                    foundOverlap = true;
                    merged = true; // Indicate that a merge happened
                }
            }
            newIntervals.push(current);
        }
        intervals = newIntervals; // Update intervals for the next pass
    } while (merged);

    // After all merges, the intervals might not be sorted, so sort them for consistent output.
    // This sorting adds O(N log N) to the overall complexity.
    // Without this final sort, the output order is unpredictable.
    intervals.sort((a, b) => a[0] - b[0]);

    return intervals;
}

// Problem 4: Trapping Rain Water (Dynamic Programming approach)

/**
 * Calculates the amount of rain water that can be trapped using Dynamic Programming.
 * This approach achieves O(N) time complexity and O(N) extra space complexity.
 *
 * The algorithm works in three passes:
 * 1. Create a `leftMax` array: `leftMax[i]` stores the maximum height of a bar to the left
 *    of or including `i`.
 * 2. Create a `rightMax` array: `rightMax[i]` stores the maximum height of a bar to the right
 *    of or including `i`.
 * 3. Iterate through the `height` array. For each bar at index `i`, the amount of water it can trap is:
 *    `min(leftMax[i], rightMax[i]) - height[i]`. Sum these values.
 *
 * @param height An array of non-negative integers representing the elevation map.
 * @returns The total amount of rain water that can be trapped.
 *
 * Time Complexity: O(N), where N is the number of bars (elements in `height`).
 *   - One pass to fill `leftMax`, one pass to fill `rightMax`, one pass to calculate water.
 * Space Complexity: O(N)
 *   - Two auxiliary arrays (`leftMax` and `rightMax`), each of size N, are created.
 */
export function trapRainWaterDP(height: number[]): number {
    const n = height.length;
    if (n <= 2) {
        return 0;
    }

    // Step 1: Create leftMax array
    const leftMax: number[] = new Array(n).fill(0);
    leftMax[0] = height[0];
    for (let i = 1; i < n; i++) {
        leftMax[i] = Math.max(leftMax[i - 1], height[i]);
    }

    // Step 2: Create rightMax array
    const rightMax: number[] = new Array(n).fill(0);
    rightMax[n - 1] = height[n - 1];
    for (let i = n - 2; i >= 0; i--) {
        rightMax[i] = Math.max(rightMax[i + 1], height[i]);
    }

    // Step 3: Calculate trapped water
    let totalWater = 0;
    for (let i = 0; i < n; i++) {
        // The amount of water trapped above bar `i` is determined by the minimum
        // of the maximum height to its left and its right, minus its own height.
        totalWater += Math.max(0, Math.min(leftMax[i], rightMax[i]) - height[i]);
    }

    return totalWater;
}
```

File: leetcode-array-processing-solutions-optimized/src/brute_force_solutions.cpp
```cpp
#include <vector>
#include <numeric> // For std::accumulate
#include <algorithm> // For std::max, std::min, std::swap, std::rotate

namespace ArrayManipulation {

// --- Brute Force / Naive solutions for comparison ---

// Problem 1: Maximum Subarray Sum
namespace MaxSubarraySum {

    /**
     * @brief Brute-force solution for Maximum Subarray Sum.
     *
     * This approach iterates through all possible subarrays, calculates their sum,
     * and keeps track of the maximum sum found.
     * A subarray is defined by its start and end indices (i, j).
     *
     * Example: nums = [-2, 1, -3, 4]
     * i=0:
     *   j=0: [-2] sum=-2. max_sum=-2
     *   j=1: [-2, 1] sum=-1. max_sum=-1
     *   j=2: [-2, 1, -3] sum=-4. max_sum=-1
     *   j=3: [-2, 1, -3, 4] sum=-0. max_sum=0
     * i=1:
     *   j=1: [1] sum=1. max_sum=1
     *   j=2: [1, -3] sum=-2. max_sum=1
     *   j=3: [1, -3, 4] sum=2. max_sum=2
     * i=2:
     *   j=2: [-3] sum=-3. max_sum=2
     *   j=3: [-3, 4] sum=1. max_sum=2
     * i=3:
     *   j=3: [4] sum=4. max_sum=4
     * Final: max_sum = 4
     *
     * @param nums The input integer array.
     * @return The maximum subarray sum.
     *
     * @complexity
     * Time: O(N^2) - Two nested loops iterate through all possible start and end points of subarrays.
     * Space: O(1) - Only a few constant extra variables.
     */
    int bruteForce(const std::vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }

        int max_so_far = nums[0];

        // Iterate through all possible starting points
        for (size_t i = 0; i < nums.size(); ++i) {
            int current_subarray_sum = 0;
            // Iterate through all possible ending points for the current starting point
            for (size_t j = i; j < nums.size(); ++j) {
                current_subarray_sum += nums[j];
                max_so_far = std::max(max_so_far, current_subarray_sum);
            }
        }

        return max_so_far;
    }

} // namespace MaxSubarraySum

// Problem 2: Product of Array Except Self
namespace ProductExceptSelf {

    /**
     * @brief Brute-force solution for Product of Array Except Self.
     *
     * For each element `nums[i]`, iterate through the entire array again
     * and multiply all elements except `nums[i]`.
     *
     * @param nums The input integer array.
     * @return An array where answer[i] is the product of all elements except nums[i].
     *
     * @complexity
     * Time: O(N^2) - For each of N elements, we iterate through N-1 other elements.
     * Space: O(N) - To store the result array.
     */
    std::vector<int> bruteForce(const std::vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return {};
        if (n == 1) return {1};

        std::vector<int> answer(n);

        for (int i = 0; i < n; ++i) {
            long long current_product = 1; // Use long long to prevent overflow during intermediate products
            for (int j = 0; j < n; ++j) {
                if (i != j) {
                    current_product *= nums[j];
                }
            }
            answer[i] = static_cast<int>(current_product);
        }
        return answer;
    }

} // namespace ProductExceptSelf

// Problem 3: Trapping Rain Water
namespace TrappingRainWater {

    /**
     * @brief Brute-force solution for Trapping Rain Water.
     *
     * For each bar `height[i]`, calculate how much water it can trap.
     * The water trapped at `height[i]` depends on the maximum height of bars to its left
     * and the maximum height of bars to its right.
     * Water trapped at `i` = `min(max_left_i, max_right_i) - height[i]`.
     * Iterate through all bars (except the first and last, which cannot trap water).
     * For each bar, scan left to find `max_left` and scan right to find `max_right`.
     *
     * @param height The elevation map.
     * @return The total amount of trapped rain water.
     *
     * @complexity
     * Time: O(N^2) - For each of N bars, we might scan left and right for N elements, roughly N + N operations.
     * Space: O(1) - Constant extra variables.
     */
    int bruteForce(const std::vector<int>& height) {
        int n = height.size();
        if (n <= 2) { // Arrays of size 0, 1, or 2 cannot trap water
            return 0;
        }

        int total_water = 0;

        for (int i = 1; i < n - 1; ++i) { // Iterate from 1 to n-2, as ends can't trap water
            int max_left = 0;
            // Find max height to the left of current bar
            for (int j = 0; j < i; ++j) {
                max_left = std::max(max_left, height[j]);
            }

            int max_right = 0;
            // Find max height to the right of current bar
            for (int j = i + 1; j < n; ++j) {
                max_right = std::max(max_right, height[j]);
            }

            // Water trapped at current bar is min(max_left, max_right) - height[i]
            // Only add if it's positive.
            total_water += std::max(0, std::min(max_left, max_right) - height[i]);
        }

        return total_water;
    }

    /**
     * @brief Dynamic Programming solution for Trapping Rain Water.
     *
     * This approach preprocesses `max_left` and `max_right` arrays to store the
     * maximum height encountered from the left up to index `i`, and from the right
     * down to index `i`.
     *
     * `left_max[i]` = max height to the left of or at `i`.
     * `right_max[i]` = max height to the right of or at `i`.
     *
     * Water trapped at `i` = `min(left_max[i], right_max[i]) - height[i]`.
     *
     * @param height The elevation map.
     * @return The total amount of trapped rain water.
     *
     * @complexity
     * Time: O(N) - Three passes: one for `left_max`, one for `right_max`, one for calculating water.
     * Space: O(N) - Two additional arrays `left_max` and `right_max` are used.
     */
    int dpSolution(const std::vector<int>& height) {
        int n = height.size();
        if (n <= 2) {
            return 0;
        }

        std::vector<int> left_max(n);
        std::vector<int> right_max(n);

        // Fill left_max array
        left_max[0] = height[0];
        for (int i = 1; i < n; ++i) {
            left_max[i] = std::max(left_max[i - 1], height[i]);
        }

        // Fill right_max array
        right_max[n - 1] = height[n - 1];
        for (int i = n - 2; i >= 0; --i) {
            right_max[i] = std::max(right_max[i + 1], height[i]);
        }

        int total_water = 0;
        // Calculate trapped water
        for (int i = 0; i < n; ++i) {
            // Water can only be trapped if min(left_max, right_max) is greater than current bar height
            total_water += std::max(0, std::min(left_max[i], right_max[i]) - height[i]);
        }

        return total_water;
    }

} // namespace TrappingRainWater

// Problem 4: Rotate Array
namespace RotateArray {

    /**
     * @brief Naive solution for Rotate Array using repeated single-element rotation.
     *
     * Rotate the array by 1 step, k times.
     * To rotate by 1 step: store the last element, shift all other elements to the right,
     * then place the stored element at the beginning.
     *
     * @param nums The integer array to be rotated (modified in-place).
     * @param k The number of steps to rotate to the right.
     *
     * @complexity
     * Time: O(N*k) - Performing k single-element rotations. Each single rotation is O(N).
     * Space: O(1) - All operations are performed in-place.
     */
    void rotateNaive(std::vector<int>& nums, int k) {
        int n = nums.size();
        if (n == 0 || k == 0) {
            return;
        }

        k %= n; // Normalize k

        for (int step = 0; step < k; ++step) {
            int last_element = nums[n - 1];
            for (int i = n - 1; i > 0; --i) {
                nums[i] = nums[i - 1];
            }
            nums[0] = last_element;
        }
    }

} // namespace RotateArray

// Problem 5: Merge Sorted Arrays
namespace MergeSortedArrays {

    /**
     * @brief Simple solution for Merge Sorted Arrays using `std::sort`.
     *
     * This method copies `nums2` into the available space in `nums1` and then
     * uses the standard library's sort function to sort the entire `nums1` array.
     *
     * @param nums1 The first array with m elements, followed by n zeros (modified in-place).
     * @param m The number of valid elements in nums1.
     * @param nums2 The second array with n elements.
     * @param n The number of valid elements in nums2.
     *
     * @complexity
     * Time: O((m+n) log(m+n)) - Copying is O(n), sorting is O((m+n) log(m+n)).
     * Space: O(1) - If `std::sort` uses an in-place sort like Introsort.
     *               However, it might use auxiliary space depending on the implementation
     *               and the specific algorithm chosen (e.g., merge sort uses O(N) space).
     *               For competitive programming, `std::sort` is often considered O(log N) or O(N) space.
     */
    void mergeUsingSort(std::vector<int>& nums1, int m, const std::vector<int>& nums2, int n) {
        // Copy elements from nums2 into the end of nums1
        for (int i = 0; i < n; ++i) {
            nums1[m + i] = nums2[i];
        }

        // Sort the entire nums1 array
        std::sort(nums1.begin(), nums1.begin() + m + n);
    }

    /**
     * @brief Solution for Merge Sorted Arrays using Auxiliary Array.
     *
     * This method creates a temporary array, merges both into it, then copies back.
     *
     * @param nums1 The first array with m elements, followed by n zeros (modified in-place).
     * @param m The number of valid elements in nums1.
     * @param nums2 The second array with n elements.
     * @param n The number of valid elements in nums2.
     *
     * @complexity
     * Time: O(m + n) - Single pass to merge, single pass to copy back.
     * Space: O(m + n) - An auxiliary array of size m+n is used.
     */
    void mergeAuxiliaryArray(std::vector<int>& nums1, int m, const std::vector<int>& nums2, int n) {
        std::vector<int> merged_array(m + n);
        int p1 = 0; // Pointer for nums1
        int p2 = 0; // Pointer for nums2
        int p_merged = 0; // Pointer for merged_array

        while (p1 < m && p2 < n) {
            if (nums1[p1] <= nums2[p2]) {
                merged_array[p_merged++] = nums1[p1++];
            } else {
                merged_array[p_merged++] = nums2[p2++];
            }
        }

        // Copy remaining elements from nums1
        while (p1 < m) {
            merged_array[p_merged++] = nums1[p1++];
        }

        // Copy remaining elements from nums2
        while (p2 < n) {
            merged_array[p_merged++] = nums2[p2++];
        }

        // Copy merged_array back to nums1
        for (int i = 0; i < m + n; ++i) {
            nums1[i] = merged_array[i];
        }
    }


} // namespace MergeSortedArrays

} // namespace ArrayManipulation
```